fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/AsyncLockRegistry.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * AsyncLockRegistry.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record;
+
+import com.apple.foundationdb.record.util.pair.ImmutablePair;
+import com.apple.foundationdb.record.util.pair.Pair;
+import com.apple.foundationdb.subspace.Subspace;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * The {@link AsyncLockRegistry} keeps track of locking over resources in the context of a single transaction. In
+ * essence, it maps a resource (by means of a {@link LockIdentifier}) to a data-structure that keeps account of
+ * outstanding read and write tasks on that resource.
+ *
+ * The consumer interacts with the registry through {@link AsyncLockRegistry#getReadLock} and
+ * {@link AsyncLockRegistry#getWriteLock} that schedule tasks and return wait {@link CompletableFuture}s. These methods
+ * take in a resource identifier {@link LockIdentifier} and a taskLock.The taskLock is a {@link CompletableFuture} that
+ * is required to be marked complete by the caller when the task is done to "leave" the lock. The methods also return a
+ * waitLock {@link CompletableFuture} that itself get marked complete when all dependant backlog tasks for the resource
+ * has completed.
+ *
+ * Hence, the caller can schedule a task as:
+ * <pre>{@code
+ *      final var taskLock = new CompletableFuture<Void>();
+ *      registry.getWriteLock(identifier, taskLock)
+ *          .thenComposeAsync(ignore -> {
+ *              // do your task...
+ *          })
+ *          .thenRun(() -> taskLock.complete(null));
```

Comment:
```suggestion
 *          .whenComplete((ignore, error) -> taskLock.complete(null));
```

We may want to add more scare text to warn people about making sure that they complete their `taskLock`s